### ISSUE TITLE
Unable to install @storybook/angular to a new Angular project.

### ISSUE THREAD

**Issue by martinsik** [storybookjs/storybook#6804]
**Describe the bug**
Running `npm install @storybook/angular --save-dev` or `npx -p @storybook/cli sb init --type angular` in a new Angular application always fails.

**To Reproduce**
Create a new Angular application using angular-cli and try to install `@storybook/angular`:

```
$ ng new test-storybook
$ cd test-storybook
$ npm install @storybook/angular --save-dev

npm WARN deprecated core-js@1.2.7: core-js@<2.6.5 is no longer maintained. Please, upgrade to core-js@3 or at least to actual version of core-js@2.
...
@babel/core@^7.0.0-0 but none is installed. You must install peer dependencies yourself.
npm WARN @babel/plugin-transform-unicode-regex@7.4.4 requires a peer of @babel/core@^7.0.0-0 but none is installed. You must install peer dependencies yourself.

npm ERR! path C:\Users\user\git\test-storybook\node_modules\array.prototype.flat
npm ERR! code ENOENT
npm ERR! errno -4058
npm ERR! syscall rename
npm ERR! enoent ENOENT: no such file or directory, rename 'C:\Users\user\git\test-storybook\node_modules\array.prototype.flat' -> 'C:\Users\user\git\test-storybook\node_modules\.array.prototype.flat.DELETE'
npm ERR! enoent This is related to npm not being able to find a file.
npm ERR! enoent

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\user\AppData\Roaming\npm-cache\_logs\2019-05-16T19_55_34_727Z-debug.log
```

**Expected behavior**
`npm i` should not fail and should install `@storybook/angular`.

**Code snippets**
`package.json` generated by angular-cli:

```
{
  "name": "test-storybook",
  "version": "0.0.0",
  "scripts": {
    "ng": "ng",
    "start": "ng serve",
    "build": "ng build",
    "test": "ng test",
    "lint": "ng lint",
    "e2e": "ng e2e"
  },
  "private": true,
  "dependencies": {
    "@angular/animations": "~7.2.0",
    "@angular/common": "~7.2.0",
    "@angular/compiler": "~7.2.0",
    "@angular/core": "~7.2.0",
    "@angular/forms": "~7.2.0",
    "@angular/platform-browser": "~7.2.0",
    "@angular/platform-browser-dynamic": "~7.2.0",
    "@angular/router": "~7.2.0",
    "core-js": "^2.5.4",
    "rxjs": "~6.3.3",
    "tslib": "^1.9.0",
    "zone.js": "~0.8.26"
  },
  "devDependencies": {
    "@angular-devkit/build-angular": "~0.13.0",
    "@angular/cli": "~7.3.8",
    "@angular/compiler-cli": "~7.2.0",
    "@angular/language-service": "~7.2.0",
    "@types/node": "~8.9.4",
    "@types/jasmine": "~2.8.8",
    "@types/jasminewd2": "~2.0.3",
    "codelyzer": "~4.5.0",
    "jasmine-core": "~2.99.1",
    "jasmine-spec-reporter": "~4.2.1",
    "karma": "~4.0.0",
    "karma-chrome-launcher": "~2.2.0",
    "karma-coverage-istanbul-reporter": "~2.0.1",
    "karma-jasmine": "~1.1.2",
    "karma-jasmine-html-reporter": "^0.2.2",
    "protractor": "~5.4.0",
    "ts-node": "~7.0.0",
    "tslint": "~5.11.0",
    "typescript": "~3.2.2"
  }
}
```

**System:**
 - OS: Windows10
 - Device: ThinkPad T480s
 - Browser: N/A
 - Framework: angular
 - Addons: N/A
 - Version: latest

**Additional context**
Node version: v11.15.0
NPM version: 6.7.0
The error was the same with Node 10.15.3.
I also tried different versions of `@storybook/angular` including `next` but the result is always the same.

Angular CLI version:
```
Angular CLI: 7.3.9
Node: 11.15.0
OS: win32 x64
Angular: 7.2.15
... animations, common, compiler, compiler-cli, core, elements
... forms, language-service, platform-browser
... platform-browser-dynamic, router

Package                            Version
------------------------------------------------------------
@angular-devkit/architect          0.13.9
@angular-devkit/build-angular      0.13.9
@angular-devkit/build-ng-packagr   0.13.9
@angular-devkit/build-optimizer    0.13.9
@angular-devkit/build-webpack      0.13.9
@angular-devkit/core               7.3.9
@angular-devkit/schematics         7.3.9
@angular/cdk                       7.3.7
@angular/cli                       7.3.9
@angular/material                  7.3.7
@ngtools/json-schema               1.1.0
@ngtools/webpack                   7.3.9
@schematics/angular                7.3.9
@schematics/update                 0.13.9
ng-packagr                         4.7.1
rxjs                               6.5.2
typescript                         3.2.4
webpack                            4.29.0
```

**Comment by arjunkalburgi** [comment]
I'm having a similar problem with my Ionic Angular project: 

When I run: `npx -p @storybook/cli sb init --type angular`
```TypeError: Cannot read property 'undefined' of undefined
    at getAngularAppTsConfigPath (/Users/smilinknight/.npm/_npx/5404/lib/node_modules/@storybook/cli/lib/helpers.js:78:24)
    at getAngularAppTsConfigPath (/Users/smilinknight/.npm/_npx/5404/lib/node_modules/@storybook/cli/lib/helpers.js:87:24)
    at editAngularAppTsConfig (/Users/smilinknight/.npm/_npx/5404/lib/node_modules/@storybook/cli/generators/ANGULAR/index.js:47:24)
    at editAngularAppTsConfig (/Users/smilinknight/.npm/_npx/5404/lib/node_modules/@storybook/cli/generators/ANGULAR/index.js:66:3)
    at call (/Users/smilinknight/.npm/_npx/5404/lib/node_modules/@storybook/cli/node_modules/regenerator-runtime/runtime.js:62:40)
    at Generator.tryCatch [as _invoke] (/Users/smilinknight/.npm/_npx/5404/lib/node_modules/@storybook/cli/node_modules/regenerator-runtime/runtime.js:288:22)
    at Generator._invoke [as next] (/Users/smilinknight/.npm/_npx/5404/lib/node_modules/@storybook/cli/node_modules/regenerator-runtime/runtime.js:114:21)
    at asyncGeneratorStep (/Users/smilinknight/.npm/_npx/5404/lib/node_modules/@storybook/cli/generators/ANGULAR/index.js:50:103)
    at _next (/Users/smilinknight/.npm/_npx/5404/lib/node_modules/@storybook/cli/generators/ANGULAR/index.js:52:194)
    at processTicksAndRejections (internal/process/task_queues.js:86:5)```

And then when I run `npm run storybook` ignoring this error I get: 
```
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! titan96@0.0.1 storybook: `start-storybook -p 6006`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the titan96@0.0.1 storybook script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/smilinknight/.npm/_logs/2019-05-18T18_08_26_409Z-debug.log
```




I did the manual setup instead but also get errors when running `npm run storybook`
```
 10% building 1/1 modules 0 active(node:4800) DeprecationWarning: Tapable.plugin is deprecated. Use new API on `.hooks` instead
(node:4800) DeprecationWarning: Tapable.apply is deprecated. Call apply on the plugin directly instead
/Users/smilinknight/projects/Titan/Titan96/node_modules/webpack/lib/SetVarMainTemplatePlugin.js:17
		compilation.templatesPlugin("render-with-entry", (source, chunk, hash) => {
		            ^

TypeError: compilation.templatesPlugin is not a function
    at SetVarMainTemplatePlugin.apply (/Users/smilinknight/projects/Titan/Titan96/node_modules/webpack/lib/SetVarMainTemplatePlugin.js:17:15)
    at Compilation.apply (/Users/smilinknight/projects/Titan/Titan96/node_modules/@storybook/core/node_modules/tapable/lib/Tapable.js:79:16)
    at Compilation.deprecated [as apply] (internal/util.js:71:15)
    at compiler.plugin (/Users/smilinknight/projects/Titan/Titan96/node_modules/webpack/lib/LibraryTemplatePlugin.js:45:18)
    at SyncHook.eval [as call] (eval at create (/Users/smilinknight/projects/Titan/Titan96/node_modules/@storybook/core/node_modules/tapable/lib/HookCodeFactory.js:19:10), <anonymous>:11:1)
    at SyncHook.lazyCompileHook (/Users/smilinknight/projects/Titan/Titan96/node_modules/@storybook/core/node_modules/tapable/lib/Hook.js:154:20)
    at Compiler.newCompilation (/Users/smilinknight/projects/Titan/Titan96/node_modules/@storybook/core/node_modules/webpack/lib/Compiler.js:580:30)
    at hooks.beforeCompile.callAsync.err (/Users/smilinknight/projects/Titan/Titan96/node_modules/@storybook/core/node_modules/webpack/lib/Compiler.js:617:29)
    at _err0 (eval at create (/Users/smilinknight/projects/Titan/Titan96/node_modules/@storybook/core/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:11:1)
    at compiler.inputFileSystem.readFile (/Users/smilinknight/projects/Titan/Titan96/node_modules/@storybook/core/node_modules/webpack/lib/DllReferencePlugin.js:72:15)
    at process.nextTick (/Users/smilinknight/projects/Titan/Titan96/node_modules/@storybook/core/node_modules/enhanced-resolve/lib/CachedInputFileSystem.js:73:15)
    at processTicksAndRejections (internal/process/task_queues.js:79:9)

```

**Comment by lslocum** [comment]
I think the problem is the new version of zone.js in Angular 8. I get this error when I try to upgrade my existing project from 7 to 8

![image](https://user-images.githubusercontent.com/19716694/58561481-ce01cc00-81ec-11e9-8de7-82e15458600d.png)

**Comment by Jason-Michaels** [comment]
The zone.js dependency error is also stopping me from using storybook on Angular 8... I hope it get's resolved soon!

**Comment by kroeder** [comment]
I couldn't reproduce it

```
npm i -g @storybook/cli @angular/cli
ng new angular-storybook-test
cd angular-storybook-test
sb init
npm run storybook

╭────────────────────────────────────────────────────╮
│                                                    │
│   Storybook 5.1.0-rc.3 started                     │
│   14 s for manager and 15 s for preview            │
│                                                    │
│   Local:            http://localhost:6006/         │
│   On your network:  http://192.168.178.20:6006/    │
│                                                    │
╰────────────────────────────────────────────────────╯
```

## Storybook version
5.1.0-rc.3

## Angular version
```
$ ng version

     _                      _                 ____ _     ___
    / \   _ __   __ _ _   _| | __ _ _ __     / ___| |   |_ _|
   / △ \ | '_ \ / _` | | | | |/ _` | '__|   | |   | |    | |
  / ___ \| | | | (_| | |_| | | (_| | |      | |___| |___ | |
 /_/   \_\_| |_|\__, |\__,_|_|\__,_|_|       \____|_____|___|
                |___/
    

Angular CLI: 8.0.1
Node: 10.15.3
OS: win32 x64
Angular: 8.0.0
... animations, common, compiler, compiler-cli, core, forms
... language-service, platform-browser, platform-browser-dynamic
... router

Package                           Version
-----------------------------------------------------------
@angular-devkit/architect         0.800.1
@angular-devkit/build-angular     0.800.1
@angular-devkit/build-optimizer   0.800.1
@angular-devkit/build-webpack     0.800.1
@angular-devkit/core              8.0.1
@angular-devkit/schematics        8.0.1
@angular/cli                      8.0.1
@ngtools/webpack                  8.0.1
@schematics/angular               8.0.1
@schematics/update                0.800.1
rxjs                              6.4.0
typescript                        3.4.5
webpack                           4.30.0
```

I'm going to upgrade our companies storybook in the next couple of days and check if it fails
The only bug I encountered was https://github.com/storybookjs/storybook/issues/6934

**Comment by brianmcd** [comment]
@kroeder - When I run those commands, I get a warning about `@angular/storybook` zone.js dependency:

`warning " > @storybook/angular@5.0.11" has incorrect peer dependency "zone.js@^0.8.27".`

Things proceed and install, but a fresh install is different than upgrading with the Angular CLI.

This fails:

```
yarn global add @angular/cli@^7 @storybook/cli
ng new ng8-upgrade-sb
sb init
ng update @angular/cli @angular/core
```

Output:
```
Package "@storybook/angular" has an incompatible peer dependency to "zone.js" (requires "^0.8.27", would install "0.9.1").
Incompatible peer dependencies found. See above.
```

And the CLI won't proceed.

I'd imagine that the zone.js `peerDependency` just needs to be bumped: https://github.com/storybookjs/storybook/blob/8fef8f408f52105f4c48a13d61ab2df2e068080e/app/angular/package.json#L54

A workaround is: 
```
ng update @angular/cli @angular/core --force
```

**Comment by martinsik** [comment]
I just tried again what I wrote in the initial post after upgrading to Angular 8 but I'm still having the exact same issue:

```
C:\Users\user\test-storybook\node_modules\array.prototype.flat
npm ERR! code ENOENT
npm ERR! errno -4058
npm ERR! syscall rename
npm ERR! enoent ENOENT: no such file or directory, rename 'C:\Users\user\test-storybook\node_modules\array.prototype.flat' -> 'C:\Users\user\test-storybook\node_modules\.array.prototype.flat.DELETE'
npm ERR! enoent This is related to npm not being able to find a file.
```

**Comment by kroeder** [comment]
A short wrap-up: There seem to be 2 individual bugs here 

1. `sb init` in an Angular 7 project and then `yarn storybook` won't work. Angular 7 depends on `core-js@^2.0.0`, Angular 7 on `core-js@^3.0.0` due to babel. There's a webpack plugin that was meant to solve that version mismatch but it seems that there's still an issue. If you got stuck with something like `Module not found: Error: Can't resolve 'core-js/modules/es.array.for-each'` then I suggest to first update to Angular 8 if possible. They dropped their dependency on `core-js`

2. If you run into a zone.js issue then it is caused by a missing version range for `zone.js` in `@storybook/angular`. It should be easy to fix - see #6951. @brianmcd thanks for opening a new issue for that. As workaround I suggest to try `ng update @angular/core --force`. Please comment in the other issue if this is the error you are running into

@arjunkalburgi Your issue in your first snippet was solved by https://github.com/storybookjs/storybook/issues/6934. If you still run into your second error after using the latest storybook RC then please open a separate issue. It doesn't seem like this is related to this issue. Correct me if I'm wrong 🙂

**Comment by arjunkalburgi** [comment]
I fixed it by upgrading to Ionic 4 and Angular 7. Which now makes sense why that worked because it upgraded those other dependencies too. Nice!

**Comment by shilman** [comment]
Is this still an issue @kroeder ?

**Comment by lslocum** [comment]
I force upgraded to Angular 8 and had no issues with storybook.

**Comment by asinitson** [comment]
Seems to be broken again somewhere **after** Angular `8.0.1`.

Steps to reproduce:

```bash
VERSION=8.2.0
PROJECT=angular-eight-two-zero
npx -p @angular/cli@$VERSION ng new --defaults=true $PROJECT
cd $PROJECT
npx -p @storybook/cli sb init --type angular
```

After this running `npm start` and `npm run storybook` produce this:
![image](https://user-images.githubusercontent.com/13285303/62461439-7607b880-b78d-11e9-93d9-d0ea1fdf5713.png)

If you substitute first two envrionment variables to:
```bash
VERSION=8.0.1
PROJECT=angular-eight-zero-one
```
...then you get to the happy end:
![image](https://user-images.githubusercontent.com/13285303/62461705-2b3a7080-b78e-11e9-884c-c62bb1dcd2cf.png)

Not sure what to do with this one. I really wanted to try the tool out, but I don't feel like downgrading my Angular project. Any workaround for Angular `8.2.0`?

**Comment by kroeder** [comment]
I'll take a look at this later this day
Seems like a dependency injection issue

can't resolve all parameters for AppComponent(?, ?) usually  means the DI Tokens couldn't be injected into AppComponent

https://github.com/storybookjs/storybook/blob/next/app/angular/src/client/preview/angular/components/app.component.ts#L38

```typescript
  constructor(
    private cfr: ComponentFactoryResolver,
    private changeDetectorRef: ChangeDetectorRef,
    @Inject(STORY) private data: Observable<StoryFnAngularReturnType>
  ) {}
```

the first two can't be resolved, but the Story Token seems to work 🤔

**Comment by kroeder** [comment]
Can confirm the above error.

This seems to be fixed in `@next`
```
npx -p @storybook/cli@next sb init --type angular
```

although there's another error in `next` 
```
ERROR in D:/04_Development/ng-projects/sb-ngcli820/node_modules/@storybook/angular/dist/client/preview/index.d.ts
ERROR in D:/04_Development/ng-projects/sb-ngcli820/node_modules/@storybook/angular/dist/client/preview/index.d.ts(2,23):
TS2688: Cannot find type definition file for 'webpack-env'.
```

A workaround for `next` is to set `skibLibCheck: true` in `.storybook/tsconfig.json`
We need to figure out the changes in `5.2.0` that might have solved this issue

We also need to figure out why `webpack-env` makes trouble in `5.2.0-beta`

**Comment by bzkny** [comment]
This workaround didn't work for me added: `"skibLibCheck": true,` to `.storybook/tsconfig.json` + npm run storybook but same message

**Comment by MrBugatti** [comment]
> Can confirm the above error.
> 
> This seems to be fixed in `@next`
> 
> ```
> npx -p @storybook/cli@next sb init --type angular
> ```

This command is all I needed to get working for Angular 8.2.0.

**Comment by shilman** [comment]
@kroeder can we close this?

**Comment by nategreen** [comment]
@bzkny Maybe you already caught this, but:

- `skibLibCheck` evidently had a typo and should've been `skipLibCheck`
- `skipLibCheck` goes in the `compilerOptions` not directly in the top level of `tsconfig.json`

I put `"skipLibCheck": true` in `compilerOptions` array in tsconfig.json and the build stopped failing.

**Comment by asinitson** [comment]
Did anyone get this to work? By work I mean did anyone manage to **do something useful with the tool beyond compiling a sample**?

This is what I did:

1. Initialized the latest versions (Angular 8.2.4, Storybook RC.5.2.0-rc.5 at this point):

```bash
PROJECT=angular-storybook-love-story
npx -p @angular/cli@ ng new --defaults=true $PROJECT
cd $PROJECT
npx -p @storybook/cli@next sb init --type angular
```
2. Added `"skipLibCheck": true` to `.storybook/tsconfig.json`

```json
{
  "extends": "../tsconfig.app.json",
  "compilerOptions": {
    "skipLibCheck": true,
```
3. Autogenerate dummy component without any `@Input`/`@Output`

```
node_modules/.bin/ng g component dummy
```
4. Create story with a component

```
import { storiesOf } from '@storybook/angular';
import { DummyComponent } from 'src/app/dummy/dummy.component';

storiesOf('Dummy `y u no work` component', module)
  .add('generated with `ng g component dummy`', () => ({
    component: DummyComponent,
    props: { }
  }));
```
5. This is the sad result I get (Chrome and Firefox):

![image](https://user-images.githubusercontent.com/13285303/64252790-7f1ead80-cf24-11e9-85b4-4c56ebeac5d7.png)

This is what seems to be the most meaningful part of the error message:
```
vendors~main.a5582890b64b0d655bfc.bundle.js:170749 Unhandled Promise rejection: Expected 'styles' to be an array of strings. ; Zone: <root> ;
```

**Any hints how to get further?** I would really like to start examining my components, but each time I try Storybook with Angular I cannot get far.

**Comment by marchino21** [comment]
Hello, I've got same issue as @asinitson and also if I remove styleUrls for component I get 
The template specified for componen is not a string

**Comment by guillermoarellano** [comment]
@asinitson @marchino21 I recently upgraded my version of Angular and it broke Storybook. I would recommend you downgrade your CLI version to 8.2.0 and the devkit 0.802.0. That fixed it for me. See this comment -- https://github.com/storybookjs/storybook/issues/7877#issuecomment-525833864

**Comment by asinitson** [comment]
@guillermoarellano : Thanks! That did the trick!

For other reading this, here are exact command to downgrade:
```bash
npm install @angular/cli@8.2
npm install @angular-devkit/build-angular@0.802
```

...which currently results in this change in package.json
```json
  "devDependencies": {
    "@angular-devkit/build-angular": "^0.802.2",
    "@angular/cli": "^8.2.2",
```

And this happy picture:
![image](https://user-images.githubusercontent.com/13285303/64506404-2e220700-d2e0-11e9-8621-757e56749b30.png)

**Comment by kroeder** [comment]
This was fixed in 5.2.4 and should already work as intended. I tested it with angular 8.2.14 and storybook 5.2 and 5.3-beta.2

I close this since everything seems to work now.
Please leave a comment if it is not fixed for your project!

**Comment by thiagosbrito** [comment]
Hello, 

I'm facing an issue while trying to add storybook to an existing project.

The project runs:

```
"@angular-devkit/build-angular": "~0.901.9",
    "@angular-devkit/build-ng-packagr": "~0.901.9",
    "@angular/animations": "~9.1.11",
    "@angular/cli": "~9.1.9",
    "@angular/common": "~9.1.11",
    "@angular/compiler": "~9.1.11",
    "@angular/compiler-cli": "~9.1.11",
    "@angular/core": "~9.1.11",
    "@angular/forms": "~9.1.11",
    "@angular/language-service": "~9.1.11",
    "@angular/platform-browser": "~9.1.11",
    "@angular/platform-browser-dynamic": "~9.1.11",
    "@angular/router": "^9.1.11",
    "@ngrx/effects": "~8.6.0",
    "@ngrx/store": "~8.6.0",
    "@types/express": "^4.17.1",
    "@types/jasmine": "~2.8.8",
    "@types/jasminewd2": "~2.0.3",
    "@types/node": "^12.11.1",
 ```
 
 Trying to install using: `npx sb init` but I'm getting the following error:
 
 ```
 $ npx -p @storybook/cli sb init --type angular
####......] - postinstall: info lifecycle @storybook/cli@6.3.2~postinstall: @storybook/cli@6.3.2pes@7.14.5
 sb init - the simplest way to add a Storybook to your project. 

 • Installing Storybook for user specified project type. ✓
 • Adding Storybook support to your "Angular" app
     TypeError: Invalid Version: undefined
    at new SemVer (C:\Users\UX159567\AppData\Roaming\npm-cache\_npx\47868\node_modules\@storybook\cli\node_modules\@storybook\semver\classes\semver.js:72:13)
    at compare (C:\Users\UX159567\AppData\Roaming\npm-cache\_npx\47868\node_modules\@storybook\cli\node_modules\@storybook\semver\functions\compare.js:6:10)
    at Object.gte (C:\Users\UX159567\AppData\Roaming\npm-cache\_npx\47868\node_modules\@storybook\cli\node_modules\@storybook\semver\functions\gte.js:6:10)
    at generator (C:\Users\UX159567\AppData\Roaming\npm-cache\_npx\47868\node_modules\@storybook\cli\dist\cjs\generators\ANGULAR\index.js:53:38)
    at runGenerator (C:\Users\UX159567\AppData\Roaming\npm-cache\_npx\47868\node_modules\@storybook\cli\dist\cjs\initiate.js:169:37)
    at installStorybook (C:\Users\UX159567\AppData\Roaming\npm-cache\_npx\47868\node_modules\@storybook\cli\dist\cjs\initiate.js:223:10)
    at _default (C:\Users\UX159567\AppData\Roaming\npm-cache\_npx\47868\node_modules\@storybook\cli\dist\cjs\initiate.js:296:10)
    at Command.<anonymous> (C:\Users\UX159567\AppData\Roaming\npm-cache\_npx\47868\node_modules\@storybook\cli\dist\cjs\generate.js:38:696)
    at Command.listener [as _actionHandler] (C:\Users\UX159567\AppData\Roaming\npm-cache\_npx\47868\node_modules\@storybook\cli\node_modules\commander\index.js:426:31)
    at Command._parseCommand (C:\Users\UX159567\AppData\Roaming\npm-cache\_npx\47868\node_modules\@storybook\cli\node_modules\commander\index.js:1002:14) 
    ```
